metadata-builders/src/test/java/com/linkedin/metadata/builders/search/DatasetIndexBuilderTest.java

```diff
@@ -14,67 +14,119 @@
 import com.linkedin.common.urn.DatasetUrn;
 import com.linkedin.data.template.StringArray;
+import com.linkedin.dataplatform.DataPlatformInfo;
+import com.linkedin.dataplatform.PlatformType;
 import com.linkedin.dataset.DatasetDeprecation;
 import com.linkedin.dataset.DatasetProperties;
 import com.linkedin.dataset.Upstream;
 import com.linkedin.dataset.UpstreamArray;
 import com.linkedin.dataset.UpstreamLineage;
+import com.linkedin.metadata.aspect.DataPlatformAspect;
 import com.linkedin.metadata.aspect.DatasetAspect;
+import com.linkedin.metadata.dao.RestliRemoteDAO;
 import com.linkedin.metadata.dao.utils.ModelUtils;
 import com.linkedin.metadata.search.DatasetDocument;
+import com.linkedin.metadata.snapshot.DataPlatformSnapshot;
 import com.linkedin.metadata.snapshot.DatasetSnapshot;
+import com.linkedin.restli.client.RestClient;
 import com.linkedin.schema.BooleanType;
 import com.linkedin.schema.SchemaField;
 import com.linkedin.schema.SchemaFieldArray;
 import com.linkedin.schema.SchemaFieldDataType;
 import com.linkedin.schema.SchemaMetadata;
 import java.util.Collections;
 import java.util.List;
+import java.util.Optional;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import org.mockito.internal.util.reflection.FieldSetter;
+import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.Test;
 
 import static org.assertj.core.api.Assertions.*;
 
 
 public class DatasetIndexBuilderTest {
-  private final DatasetIndexBuilder _indexBuilder = new DatasetIndexBuilder();
+
+  private static final String DATASET_BAR = "/bar";
+  private static final String PLATFORM_FOO = "foo";
+
+  @Mock
+  private RestClient mockRestClient;
+  @Mock
+  private RestliRemoteDAO<DataPlatformSnapshot, DataPlatformAspect, DataPlatformUrn> dataPlatformDAO;
+
+  private DatasetIndexBuilder indexBuilder;
+
+  @BeforeMethod
+  public void setup() throws NoSuchFieldException {
+    MockitoAnnotations.initMocks(this);
+    Mockito.when(dataPlatformDAO.get(Mockito.eq(DataPlatformInfo.class), Mockito.eq(new DataPlatformUrn("hdfs"))))
+        .thenReturn(Optional.of(
+            new DataPlatformInfo().setName("hdfs").setType(PlatformType.FILE_SYSTEM).setDatasetNameDelimiter("/")));
+
+    Mockito.when(dataPlatformDAO.get(Mockito.eq(DataPlatformInfo.class), Mockito.eq(new DataPlatformUrn("foo"))))
+        .thenReturn(Optional.of(
+            new DataPlatformInfo().setName("foo").setType(PlatformType.FILE_SYSTEM).setDatasetNameDelimiter("/")));
+
+    Mockito.when(dataPlatformDAO.get(Mockito.eq(DataPlatformInfo.class), Mockito.eq(new DataPlatformUrn("hive"))))
+        .thenReturn(Optional.of(
+            new DataPlatformInfo().setName("hive").setType(PlatformType.FILE_SYSTEM).setDatasetNameDelimiter(".")));
+
+    indexBuilder = new DatasetIndexBuilder(mockRestClient);
+
+    FieldSetter.setField(indexBuilder, indexBuilder.getClass().getDeclaredField("dataPlatformDAO"), dataPlatformDAO);
```

Comment:
All this mocking and setting of private fields indicates your API here is probably in need of a refactor. These are brittle practices and antipatterns.

- If the field name is ever changed or the field removed, or made final, tests will fail rather than a compile time error. _Private fields are not part of an API and should not be access via reflection._ Generally speaking _avoid reflection_. There's usually a more elegant way of expressing what you want without it, and in a less brittle way too.
- go/lessmocks https://testing.googleblog.com/2013/05/testing-on-toilet-dont-overuse-mocks.html


Suggestions:

- Create a small interface that, given a data platform urn, returns the delimiter.
- Create one implementation of that interface that uses restli to get it. Move the cache here too.
- In tests you can use a fake implementation, even just a lambda...
- Have `DatasetIndexBuilder` take the interface in its constructor. You can even have multiple constructors if you want, one that takes this interface (for tests) and one that takes a RestClient and makes the rest implementation above (for production).

No more mocks or awkward setting of private fields. Also means private fields can be `final`.

tl;dr

```
interface DatasetPlatformPaths {
  String getDelimiter(DataPlatformUrn urn);
}
```

```
final class RestliDatasetPlatformPaths implements DatasetPlatformPaths {
  // implementation goes here, include the cache here, etc
  RestliDatasetPlatformPaths(Client client) {...}
}
```

```
class DatasetIndexBuilder {
  // Visible for testing
  DatasetIndexBuilder(@Nonnull DatasetPlatformPaths restliClient) {
    ...
  }
  public DatasetIndexBuilder(@Nonnull Client restliClient) {
    this(new RestliDatasetPlatformPaths(client));
  }
}
```

And now tests become:

```
indexBuilder = new DatasetIndexBuilder((urn) => {
  // ... return "." if hive, "/" if foo, etc
});
```